features/steps/givens/attributes.py

```diff
@@ -63,18 +64,13 @@ def step_impl(context, file_or_model, field, values):
     context.applicable = getattr(context, 'applicable', True) and applicable
 
 
-@given('Its attribute {attribute}')
+@validate_step('Its attribute {attribute}')
 def step_impl(context, attribute):
     context._push()
     context.instances = misc.map_state(context.instances, lambda i: getattr(i, attribute, None))
     setattr(context, 'attribute', attribute)
 
 
-@given('The {representation_id} shape representation has RepresentationType "{representation_type}"')
-def step_impl(context, representation_id, representation_type):
-    context.instances = list(filter(None, list(map(lambda i: ifc.instance_getter(i, representation_id, representation_type), context.instances))))
-
-
-@given("An IFC model")
+@validate_step("An IFC model")
```

Comment:
Yes, in the IFC rule: https://github.com/buildingSMART/ifc-gherkin-rules/blob/main/features/IFC001-only-official-ifc-versions-allowed.feature#L11

```
@implementer-agreement
@critical
@IFC
Feature: IFC101 - Only Official IFC versions allowed

This rule verifies that the IFC model has a schema identifier corresponding to any of the official versions released by buildingSMART.
Specifically, IFC2x3 TC1 (version 2.3.0.1), IFC4 ADD2 TC1 (version 4.0.2.1), or IFC4.3 ADD2 (version 4.3.2.0).

  Scenario: Verifying Current Schema Identifier for IFC version
  
    Given An IFC model
    Then The Schema Identifier of the model must be "IFC4X3_ADD2" or "IFC4" or "IFC2X3"
```

Reply:
Ah ok, thanks. I hadn't seen that yet. It's a bit of a no-op though since the fact that we're operating on an IFC model is implicitly implied in all other scenarios. Even if it reads pleasantly, a `given` statement is not required. I'd propose to remove the Given statement here and keep only the then statement. If you don't like `then` without any antecedent, a synonym for `then` is `*`. In my view the entire scenario should be `* The Schema Identifier of the model must be "IFC4X3_ADD2" or "IFC4" or "IFC2X3"`.

Reply:
The problem here is that if we have no Given statement, we don't set the applicable = True in the context. This causes the rule to basically skip the then statement and become NA. 

I don't have a good solution to this yet and lean towards asking to "save this one for later", especially since it was not added in this PR. What do you think?

Reply:
Ok, you're right, let's forget about fixing this for now. You might not be correct about it being skipped by default though, because we seem to be doing a getattr() with default=True.

https://github.com/buildingSMART/ifc-gherkin-rules/pull/119/files#diff-2c838ec4f24dc4e43b898ddaa7de4de8f6cc972099850db514753312625e105aR199

```
    if not getattr(context, 'applicable', True):
```

But I also found this added in this PR, which might the behaviour you are thinking of. I would say that context.applicable and context.instances should ideally be fully orthogonal, do you by any chance remember why this if statement was introduced?

https://github.com/buildingSMART/ifc-gherkin-rules/pull/119/files#diff-4f0aef67b4b6ecf25f1ac069a7eb8a1809f980166c756e2fdbf5ac43e8e62f04R30-R33

```
    if context.instances:
        context.applicable = getattr(context, 'applicable', True)
    else:
        context.applicable = getattr(context, 'applicable', False)
```

Reply:
I looked it up, it seems it has to do with the rule SPS001
https://github.com/buildingSMART/ifc-gherkin-rules/blob/main/features/SPS001_Basic-spatial-structure-for-buildings.feature

I think it was applying to the case in which we search for multiple instances in an IFC file. For instance 
```
Given an IfcSite
Given an IfcBuilding
```
If we don't find an `IfcSite`, but we do find an `IfcBuilding`, the 'Given' condition is not met. 
However, without this statement, the 'Then' step_impl afterwards will run with the instances returned by the 'Given an IfcBuilding'. With the statement, it won't run due to applicability being False.

An idea is to move this logic to the decorator; if an empty list is returned, set applicability to False. In theory ..

Reply:
```
Given an IfcSite
Given an IfcBuilding
```

In database speak this should give you the cartesian product pairs of all sites combined with all buildings. In our case I think we should simply not allow this. But let's park this discussion :)